openlibrary/plugins/admin/code.py

```diff
@@ -240,74 +241,71 @@ def GET(self):
+            edition = self.get_edition_by_ocaid(i.old_ocaid or ocaid)
+            if not edition:
+                raise web.badrequest('No openlibrary_edition specified in item')
 
+        if not edition:
+            edition = web.ctx.site.get(f'/books/{metadata.get("openlibrary_edition")}')
         if not edition:
             raise web.notfound()
 
-        # Update record
-        match i.get('action', ''):
-            case 'remove':
-                self.remove_ocaid(edition)
-            case 'modify':
-                self.modify_ocaid(edition, i.get('ocaid'))
-            case '_':
-                raise web.badrequest('Unknown action')
+        if metadata.get('is_dark'):
+            self.remove_ocaid(edition)
+            return delegate.RawText(json.dumps({"status": "ok"}))
 
-        return delegate.RawText(json.dumps({"status": "ok"}))
+        # XXX : Should this occur before we fetch the metadata?  Is there a
+        # need to fetch item metadata if we're modifying the ocaid?
+        if i.old_ocaid:
+            self.modify_ocaid(edition, i.old_ocaid)
+            return delegate.RawText(json.dumps({"status": "ok"}))
 
-    def is_authorized(self, access_key, secret_key):
-        """Returns True if account is authorized to make changes to records."""
-        auth = accounts.InternetArchiveAccount.s3auth(access_key, secret_key)
+        return web.badrequest('ocaid was not renamed, nor was it dark')
 
-        if not auth.get('username', ''):
-            return False
+    def bounce_authorized(self):
+        """Returns True if account is authorized to make changes to records."""
+        # Authenticate request:
+        s3_access_key = web.ctx.env.get('HTTP_X_S3_ACCESS', '')
+        s3_secret_key = web.ctx.env.get('HTTP_X_S3_SECRET', '')
+        auth = accounts.InternetArchiveAccount.s3auth(s3_access_key, s3_secret_key)
 
-        acct = accounts.OpenLibraryAccount.get(email=auth.get('username'))
-        user = acct.get_user() if acct else None
+        if auth.get('username', ''):
+            acct = accounts.OpenLibraryAccount.get(email=auth.get('username'))
+            user = acct.get_user() if acct else None
 
-        if not user or (user and not user.is_usergroup_member('/usergroup/ia')):
-            return False
+            if user and user.is_usergroup_member('/usergroup/ia'):
+                return True
 
-        return True
+        raise web.unauthorized()
 
-    def validate_input(self, i):
-        """Returns True if the request is valid.
-        All requests must have an olid and an action.  If the action is
-        'modify', the request must also include 'ocaid'.
-        """
-        action = i.get('action', '')
-        return 'olid' in i and (
-            action == 'remove' or (action == 'modify' and 'ocaid' in i)
-        )
+    def get_edition_by_ocaid(self, ocaid):
+        ocaid = ocaid.replace('_', ' ')
+        q = {'type': '/type/edition', 'ocaid': ocaid}
```

Comment:
I modeled this on the code used to display a book page when an `ocaid` is used, [here](https://github.com/internetarchive/openlibrary/blob/409914bf541b32b2160200b7623060f2b5fab6c0/openlibrary/plugins/openlibrary/code.py#L500-L513).  Not sure if the underscore replacement is still necessary these days.

Reply:
```suggestion
        q = {'type': '/type/edition', 'ocaid': ocaid}
```